Add visualization features with support to MSCC data resetting

francescoraves483 opened the pull request:
This pull request proposes a set of enhancements/additional features to CCEdit, in the hope that they may be useful for the next version of the editor.
They include, but are not limited to, a couple of advanced options to show clone buttons that may be related to data resetting in MSCC. Indeed, currently all clone button connections outside the range [0,31] are marked as invalid. However, in few cases, connections to (x,32) may be made on purpose to trigger data resetting.

A more exhaustive list of the enhancements is the following:
- visualizing clone connections to (x,32) as "`INVALID / MSCC Data Res. (<reason>)`" in the Advanced Mechanics panel, to show that they may be made on purpose to trigger data resetting on MSCC
- Possibility to view the clone buttons/clone machines numbers (in a similar way as it was possible with the monster list); they will be reported with red numbers, while the monster list entries have been modified to use blue numbers
- Possibility to view the trap buttons/traps numbers; they will be reported with brown numbers
- Possibility to enable the display of button and teleport connections when the mouse hovers over a button, trap, clone machine or teleport
- A few advanced options, valid for MSCC only and grouped under "View" -> "Advanced (MSCC Only)", including (1) marking clone buttons with connections to (x,32) as data resetting buttons, showing a small icon with the corresponding effect (e.g., resetting the Chip X-coordinate), (2) marking tiles with tanks that may trigger the Multiple Tank Glitch in MSCC

I also updated the year reported in the copyright notice, as I noticed that it was still reporting "2020".

francescoraves483 committed:
Add visualization features with support to MSCC data resetting

They include:
- visualizing clone connections to (x,32) as "INVALID / MSCC Data Res. (<reason>)" in the Advanced Mechanics panel, to show that they may be made on purpose to trigger data resetting on MSCC
- Possibility to view the clone buttons/clone machines numbers (in a similar way as it was possible with the monster list); they will be reported with red numbers, while the monster list entries have been modified to use blue numbers
- Possibility to view the trap buttons/traps numbers; they will be reported with brown numbers
- Possibility to enable the display of button and teleport connections when the mouse hovers over a button, trap, clone machine or teleport
- A few advanced options, valid for MSCC only and grouped under "View" -> "Advanced (MSCC Only)", including (1) marking clone buttons with connections to (x,32) as data resetting buttons, showing a small icon with the corresponding effect (e.g., resetting the Chip X-coordinate), (2) marking tiles with tanks that may trigger the Multiple Tank Glitch in MSCC

I also updated the year reported in the copyright notice.

zrax commented on src/CCEdit/CCEdit.cpp:
```diff
@@ -1245,6 +1293,15 @@ void CCEditMain::populateTilesets()
         for (const QString& file : path.entryList(tilesetGlob, QDir::Files | QDir::Readable, QDir::Name))
             tilesets << path.absoluteFilePath(file);
     }
+
+    // Search starting from the build directory when using CLion
+    path.setPath(QApplication::applicationDirPath());
+    for(int i = 0; i < 3; ++i) {
+        path.cdUp();
+    }
+    path.cd(QStringLiteral("share/cctools"));
+    for (const QString& file : path.entryList(tilesetGlob, QDir::Files | QDir::Readable, QDir::Name))
+        tilesets << path.absoluteFilePath(file);
```
This doesn't really belong here...  Personally, I just copy the necessary files into the build dir to make it happy for debugging.

zrax commented on .gitignore:
```diff
@@ -12,3 +12,6 @@ desktop.ini
 /build
 /res/*.tis
+.idea
+cmake-build-debug
+cmake-build-optimized
```
```suggestion
/cmake-build-*
```

zrax commented on src/CCEdit/AdvancedMechanics.cpp:
```diff
@@ -298,20 +303,89 @@ QTreeWidgetItem* AdvancedMechanicsDialog::addTrapItem(const ccl::Trap& trap)
-QTreeWidgetItem* AdvancedMechanicsDialog::addCloneItem(const ccl::Clone& clone)
-{
-    QTreeWidgetItem* item = new QTreeWidgetItem(m_cloneList);
+QTreeWidgetItem* AdvancedMechanicsDialog::addCloneItem(const ccl::Clone& clone) {
+    QTreeWidgetItem *item = new QTreeWidgetItem(m_cloneList);
```
Please avoid trivial formatting changes on otherwise unchanged lines...

zrax commented on src/CCEdit/EditorWidget.cpp:
```diff
@@ -363,6 +371,122 @@ void EditorWidget::renderTo(QPainter& painter)
+                        break;
+                }
+
+                // This if() is here just for an additional sanity check, but can be potentially removed, since
+                // num_img is never expected to assume a value < 0 or > 16
```
You can use a Q_ASSERT for that

zrax commented on .gitignore:
```diff
@@ -12,3 +12,6 @@ desktop.ini
 /bin/Debug
 /build
 /res/*.tis
+.idea
```
```suggestion
/.idea
```

francescoraves483 replied:
@zrax Thank you very much for your feedback and for your first pass on the pull request. Unfortunately in these days I'm a bit busy with my work, but I'm going to tackle all the first pass comments as soon as I'm able to!

zrax reviewed (commented):
There's a lot to look through here, and I may not be able to get to it (except my basic first pass comments) for a while...  The feature sounds like a useful addition though!

francescoraves483 commented:
@zrax Thank you very much for your feedback and for your first pass on the pull request. Unfortunately in these days I'm a bit busy with my work, but I'm going to tackle all the first pass comments as soon as I'm able to!

francescoraves483 committed:
Improve code and fix .gitignore

This commit improves the code for the visualization features with support to MSCC data resetting, based on the first pull request comments.

francescoraves483 commented:
Unfortunately it took me a bit of time, but I was finally able to work on it today, and I should have tackled the first PR comments with the last commit. Thank you for the comments!

francescoraves483 committed:
Fix the text shown when hovering the mouse over a (32,y) connection

This commit fixes the text that was previously shown when the mouse
hovers a possibly invalid connection to a (32,y) location.
Previously, the text "←?" was displayed. However, while this was
properly rendered under Linux, it was under Windows.
Thus, the text has been modified to "32,y".
